### ISSUE TITLE
Spec and designs for showing full table results in emailed pulses and alerts

### ISSUE THREAD

**Issue by mazameli** [metabase/metabase#6587]
These are the design specs for issue #3894.

**Why we would do this:**
- In many organizations, there are lots of team members and executives who want to receive regular snapshot updates of metrics by a dimension or two (i.e., a pivot table), or external customers who require regular email reports. 
- https://github.com/metabase/metabase/pull/6556 will allow CSVs or XLS file attachments, but that doesn't solve the problem of "I want to scan my daily KPI email real quickly."
- Currently, emailed pulses only show the first 3 columns of pivot tables, and only the first 10 rows. They also do not allow raw data questions to be included at all. These are two artificial limitations on pulses that bug a vocal subset of our users quite a lot.

**HTML tables in pulse/alert emails**
- How many columns do we show? 
  - Up to 10. 
- How many rows do we show? 
   - Up to 20. 
- If the table exceeds either the max number of rows or columns:
    - We render as much of the HTML table as will fit in a 10x20 table
    - We'll put a bit of text between the table heading and the table itself that says “Full results have been included as a file attachment,” and we'll attach a CSV with the full results.
- The tables will scroll horizontally as needed. I think it will need to in cases where there are more than 5 columns on a mobile device.
    - As an aside, I’m personally not perturbed by this. I think allowing for a bit more leeway in the number of permitted columns is important to allow for real-world scenarios, and I honestly don’t think scrolling horizontally in a mobile email feels too bad, especially if we’re restricting the columns to 10.

**Image tables for Slack pulses/alerts**
- Just always render the table as an image, up to a 10x20 table. In Slack, you can click the image to expand it, and pan around the table.
- If the table has more than 10 columns, or more than 20 rows, render the image including the first 10 columns and first 20 rows, and include an additional bit of text in the table image that says “Showing the first 10/20 columns/rows,” or “Showing the first 10 columns and 20 rows” if it exceeds both.
- Maybe we also append “— click for full results” to the card title in Slack?

----

**CSS**
Not the prettiest code ever, but I hope this gives the basic idea.
```
h2 {
    color:#509ee3;
    font-size:18px;
    margin-bottom:5px
}
h4 {
    color:#bcc5ca;
    padding-top:0;
    margin-top:0
}
.table-style {
    width:100%;
    white-space:nowrap;
    border-collapse:collapse;
    font-family: Helvetica, sans-serif;
    text-align: right;
    color: #394044;
    font-size:12.64px;
}
.table-style td, th {
    white-space: nowrap;
    border-bottom:1px solid #EDF0F1;
    padding: 3px 3px 3px 16px;
}
.numeric-cell {
    font-family:monospace;
    text-align: right;
}
.numeric-col-heading {
    text-align: right;
}
```

**Example**

<img width="1095" alt="screen shot 2018-02-22 at 5 23 06 pm" src="https://user-images.githubusercontent.com/2223916/36573324-29e0ec44-17f5-11e8-93e9-dd4098330339.png">

**Mobile**

![img_f38f5af95f3f-1](https://user-images.githubusercontent.com/2223916/36573373-72e4e74c-17f5-11e8-838a-456dbd38dfe0.jpeg)

**Comment by boringusername99** [comment]
This makes sense to me. You may also consider taking a page out of Looker's book. They let you select what format to receive. If you select CSV you get CSV attachments only. If you select HTML, you get inline up to a certain number of rows (similar to your 20-row proposal) and then attach an HTML file when the limit is exceeded.

This is particularly beneficial to CSV on mobile, as my hunch is that most people don't have a great way of consuming CSV files on their phones/tablets.

**Comment by makmende** [comment]
This is the single most impactful feature missing for me I'd say. I was considering shifting to Jaspersoft because of the lack of proper email reports. Could you believe it?! Yes, I'd have gone back to monolythic jaspersoft just for some of the reports that we really need to receive via email. Thanks for this. Looking forward to seeing this feature live soon.

**Comment by mokra** [comment]
This feature is very important as far as I can tell. Limiting the "preview" to a fixed amount should be up to the user. Naturally, a default number is needed, but this dessicion should be taken by the user - not the developer. So I can imagine that each pulse could have some kind of extended properties - one of which would be "Max row amount in HTML format". 

As for the CSV/XLS - I believe this sounds good. The only thing I already see coming are wishes for the Excel files to be formated (with a logo for example) - this could become a kind of licensee based offer. 

Same goes for the actual HTML-Email. People will wish to use their own templates.

**Comment by georgiraychev** [comment]
Why is this issue marked as closed? I am using 0.28.0 and I am unable to add table in a Pulse.

**Comment by mazameli** [comment]
Yeah, I think this was mistaken for an issue that referenced including csvs or xls file attachments. This is not the same thing and has not yet been implemented.

**Comment by TinLe** [comment]
@mazameli Your proposal look good.  Agreed with having a start point and iterate.  I would like have some of the limits configurable.  E.g. chose to display more than 20 rows.   You can still put a hard limit, say 50 rows max, with default of 20 rows.

Thank you for doing this.

**Comment by mazameli** [comment]
Implemented in https://github.com/metabase/metabase/pull/7022

This will be included in the next version of Metabase, 0.29.
